node/uhpo/src/payout_update.rs

```diff
@@ -0,0 +1,275 @@
+    };
+
+    let mut sighash_cache = SighashCache::new(transaction.clone());
+
+    let sighash = sighash_cache.taproot_key_spend_signature_hash(
+        input_idx,
+        &Prevouts::All(prevouts),
+        TapSighashType::All,
+    )?;
+
+    let message = Message::from_digest(sighash.as_raw_hash().to_byte_array());
+
+    let signature = secp.sign_schnorr_with_rng(&message, &keypair, &mut rand::thread_rng());
+    let mut vec_sig = signature.serialize().to_vec();
+    vec_sig.push(0x01);
+
+    secp.verify_schnorr(&signature, &message, &keypair.x_only_public_key().0)?;
+
+    transaction.input[input_idx].witness.push(vec_sig);
+
+    Ok(())
+}
+
+// unit tests
+#[cfg(test)]
+mod tests {
+    use bitcoin::{key::Keypair, secp256k1::All, Network};
+    use rand::Rng;
+
+    use super::*;
+
+    fn setup() -> (Secp256k1<All>, Keypair, Address) {
+        let secp = Secp256k1::new();
+        let mut rng = rand::thread_rng();
+
+        let data: [u8; 32] = rng.gen();
+        let keypair = SecretKey::from_slice(&data).unwrap().keypair(&secp);
+
+        let new_payout_address: Address =
+            Address::p2tr(&secp, keypair.x_only_public_key().0, None, Network::Regtest);
+
+        (secp, keypair, new_payout_address)
+    }
+
+    #[test]
+    fn test_new_payout_update_only_cb() {
+        let (_, _, new_payout_address) = setup();
+
+        let coinbase_tx = create_dummy_transaction();
+        let prev_update_tx = None;
+        let projected_fee = Amount::from_sat(1000);
+
+        let payout_update = PayoutUpdate::new(
+            prev_update_tx,
+            coinbase_tx,
+            new_payout_address,
+            projected_fee,
+        )
+        .expect("Failed to create payout update");
+
+        assert_eq!(payout_update.transaction.input.len(), 1);
+        assert_eq!(payout_update.transaction.output.len(), 1);
+        assert_eq!(
+            payout_update.transaction.output[0].value,
+            Amount::from_sat(50000000) - projected_fee
+        );
+        assert!(payout_update.prev_update_txout.is_none());
+    }
+
+    #[test]
+    fn test_new_payout_update() {
+        let (_, _, new_payout_address) = setup();
+
+        let coinbase_tx = create_dummy_transaction();
+        let prev_update_tx = create_dummy_transaction();
+        let projected_fee = Amount::from_sat(1000);
+
+        let payout_update = PayoutUpdate::new(
+            Some(prev_update_tx),
+            coinbase_tx,
+            new_payout_address,
+            projected_fee,
+        )
+        .expect("Failed to create payout update");
+
+        assert_eq!(payout_update.transaction.input.len(), 2);
+        assert_eq!(payout_update.transaction.output.len(), 1);
+        assert_eq!(
+            payout_update.transaction.output[0].value,
+            Amount::from_sat(50000000 * 2) - projected_fee
+        );
+        assert!(payout_update.prev_update_txout.is_some());
+    }
+
+    #[test]
+    fn test_add_signature() {
```

Comment:
We should add tests for the code paths where:
1. No tweak is passed
2. There is an error on line 140
3. There is an error on line 158
4. Add test where signature verification fails - you might need to use mocks for this

Reply:
With latest changes made,
- new testcases are added to cover tweak keypair case.
- mock `KeypairCreationError` and `SignatureVerificationError` using `mockall` crate and making SecretKey , Secp256k1 and Keypair Generic.